src/Libraries/Microsoft.Extensions.Http.Telemetry/Logging/Internal/ServiceProviderExtensions.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Extensions.Http.Telemetry.Logging.Internal;
+
+internal static class ServiceProviderExtensions
+{
+    public static T GetRequiredKeyedServiceOrDefault<T>(this IServiceProvider serviceProvider, string? serviceKey)
```

Comment:
"Default" usually implies "default value of T" (which is typically `null` or `0`). This API appears to be retrieving a _required keyed service if the key is available_, alternatively the return value is a _required service_. `GetRequiredService` should never return defaults, it should return a service or fail....

I can't really of a good name, but I think `GetRequiredKeyedService` is a better name than this because the "OrDefault" suffix is misleading.

Reply:
`GetRequiredKeyedService` method is already defined in runtime, so I can't use it.
Please also bear in mind that it's not a public API.

Reply:
In fact, the method semantic is `GetRequiredKeyedServiceOrNonKeyedOtherwise` which I personally found a bit too long and clumsy.

Reply:
I'm all for finding a good name for the method, so if you have any other ideas - please share.

Reply:
I find the more verbose name (e.g., `GetRequiredKeyedServiceOrNonKeyedOtherwise`) better than the "...OrDefault". Especially, since it's a non-public method the length is not a big issue.
What do you think of `GetRequiredOrKeyedService`?

Reply:
Sounds good!

Reply:
Renamed